staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -926,9 +926,21 @@ func (c *Cacher) dispatchEvents() {
 			bookmarkTimer.Reset(wait.Jitter(time.Second, 0.25))
 			// Never send a bookmark event if we did not see an event here, this is fine
 			// because we don't provide any guarantees on sending bookmarks.
+			//
+			// Just pop closed watchers and requeue others if needed.
+			//
+			// TODO(#115478): rework the following logic
+			//  in a way that would allow more
+			//  efficient cleanup of closed watchers
 			if lastProcessedResourceVersion == 0 {
-				// pop expired watchers in case there has been no update
-				c.bookmarkWatchers.popExpiredWatchers()
+				for _, watchers := range c.bookmarkWatchers.popExpiredWatchers() {
+					for _, watcher := range watchers {
+						if watcher.stopped {
+							continue
+						}
+						c.bookmarkWatchers.addWatcher(watcher)
```

Comment:
how do you feel about introducing a new function that would allow for adding an array of watchers ?

Reply:
the benefit would be acquiring the lock just once.